### DISCUSSION TITLE
Writing to UART via StreamWriter from Websocket results in 'NoneType object isn't iterable'

### DISCUSSION THREAD

**Discussion by JohannesNeu** [micropython/micropython#13050] [Libraries & Drivers] [answered]
Hi,

I try to forward data from a websocket to UART via Microdot and Asyncio StreamWriter.
The following coroutine is used:
```py
async def forward_to_uart(ws):
    swriter = asyncio.StreamWriter(uart, {})    
    while True:
        stringtowrite = await ws.receive()
        bytestowrite = stringtowrite.encode()
        if bytestowrite is None:
            print("It is None, proof: " + str(bytestowrite))
            pass
        else:
            print("It is not None, but: " + str(bytestowrite))
            await swriter.write(bytestowrite)
            await swriter.drain()
        await asyncio.sleep_ms(1)
```
This is how I start the coroutine:
```py
@app.route('/echo')
@with_websocket
async def echo(request, ws):
    asyncio.create_task(forwarded_from_uart(ws))
    asyncio.create_task(forward_to_uart(ws))    
    while True:
        await asyncio.sleep_ms(1000)

async def main():
    await app.run(debug=True)

asyncio.run(main())
```
As you can see, I have tried to filter out the Nones and make sure only bytes are forwarded. Yet, nonetheless, I am getting this error message:
```
Starting async server on 0.0.0.0:5000...
GET /xterm 200
GET /node_modules/xterm-addon-attach/lib/xterm-addon-attach.js 200
GET /node_modules/xterm-addon-fit/lib/xterm-addon-fit.js 200
GET /node_modules/xterm/css/xterm.css 200
GET /node_modules/xterm/lib/xterm.js 200
It is not None, but: b'\x1b[A'
Task exception wasn't retrieved
future: <Task> coro= <generator object 'forward_to_uart' at 3ffddd20>
Traceback (most recent call last):
  File "asyncio/core.py", line 1, in run_until_complete
  File "<stdin>", line 30, in forward_to_uart
TypeError: 'NoneType' object isn't iterable

```
Where am I missing the NoneTypes?

**Comment by skyguy** [comment, marked as answer]
I think the problem is here:
`await swriter.write(bytestowrite)`
write() isn't a coroutine, so doesn't need await.

-Kevin

**Comment by JohannesNeu** [comment]
Hi @skyguy ,
that was exactly the problem! Thank you very much for the advice. I would not have been able to figure this out on my own.
Looking at the `asyncio.sleep_ms` functions: I chose the values at random and they seem to work, but I do not know how to get to "proper" values. Are there rules of thumb? Should I put the longer sleep in the `main()` function?

**Comment by skyguy** [comment]
A lot of the time, you don't need to call sleep() at all in an 
asyncio-based program - your program will wait for input somewhere and 
sit idle until the input arrives. In your code above, ws.receive() is 
where the program would automatically pause, so you could change the 
sleep to `await asyncio.sleep_ms(0)` just to give other tasks a chance 
to run.

In programs which do need a delay (e.g. they cause the processor to run 
at 100%) the most common solution I've seen is to use `sleep(0.1)` - the 
100ms delay isn't noticeable for humans who have just clicked a button, 
but is long enough that the CPU effectively sits 'idle' most of the time.

Having said that, I'm no expert! Hopefully if I'm wrong someone who 
knows better will come along and educate us both.

-Kevin

**Comment by JohannesNeu** [comment]
Thank you for the explanation. That really helped.

**Comment by sophiedegran** [comment]
You can also have a look to https://github.com/peterhinch/micropython-async/blob/master/v3
They are lots of examples with asyncio
